audit/generic_auditor.py

```diff
@@ -0,0 +1,575 @@
+        for participant in participant_data:
+            if not participant.get('samples'):
+                logging.info(
+                    f'Participant {participant.get("id")} / {participant.get("externalId")} from {self.project} has no samples. Filtering this participant out.'
+                )
+        participant_data = [
+            participant
+            for participant in participant_data
+            if participant.get('samples')
+        ]
+
+        return participant_data
+
+    def map_participants_to_samples(self, participants: list[dict]) -> dict[str, list]:
+        """
+        Returns the {external_participant_id : sample_internal_id} mapping for a Metamist project
+        - Also reports if any participants have more than one sample
+        """
+
+        # Create mapping of participant external ID to sample ID
+        participant_eid_sample_id_map = {
+            participant.get('externalId'): [
+                sample.get('id') for sample in participant.get('samples')
+            ]
+            for participant in participants
+        }
+
+        # Create mapping of participant external ID to internal ID
+        participant_eid_to_iid = {
+            participant.get('externalId'): participant.get('id')
+            for participant in participants
+        }
+
+        # Report if any participants have more than one sample ID
+        for participant_eid, sample_ids in participant_eid_sample_id_map.items():
+            if len(sample_ids) > 1:
+                logging.info(
+                    f'Participant {participant_eid_to_iid[participant_eid]} (external ID: {participant_eid})'
+                    + f' associated with more than one sample id: {sample_ids}'
+                )
+
+        return participant_eid_sample_id_map
+
+    def map_internal_to_external_sample_ids(
+        self,
+        participants: list[dict],
+    ) -> dict[str, str]:
+        """
+        Returns the {internal sample ID : external sample ID} mapping for all participants in a Metamist project
+        Also removes any samples that are part of the exclusions list
+        """
+        # Create a list of dictionaries, each mapping a sample ID to its external ID
+        samples_all = [
+            {
+                sample.get('id'): sample.get('externalId')
+                for sample in participant.get('samples')
+            }
+            for participant in participants
+        ]
+
+        # Squish them into a single dictionary not a list of dictionaries
+        sample_internal_external_id_map = {
+            sample_id: sample_external_id
+            for sample_map in samples_all
+            for sample_id, sample_external_id in sample_map.items()
+            if sample_id not in EXCLUDED_SAMPLES
+        }
+
+        return sample_internal_external_id_map
+
+    def get_sequence_mapping(
+        self, all_sequences: list[dict]
+    ) -> tuple[dict[int, str], defaultdict[int, list]]:
```

Comment:
(This is slightly mis-annotated, you're adding tuples, not a list)
One thing you can do for clarity, is define some placeholder for IDs, eg:

```python
SequenceId = int
SampleId = str
SampleExternalId = str
# ... etc

def get_sequence_mapping(all_sequences: list[dict]) -> tuple[dict[SequenceId, SampleId], dict[SequenceId, tuple[str, int]]):